### ISSUE TITLE
jQuery MultiSelect UI Widget 1.13 by Eric Hynds does not work with free-jqgrid 4.13.3

### ISSUE THREAD

**Issue by AshK14** [free-jqgrid/jqGrid#252]
Hi Oleg,

I cannot get the MultiSelect UI Widget to work with free-jqgrid 4.13.3. In a column with multiselect filter, we can select upto 2 values, the grid is filtered, but when 3rd value is selected, the filter is reset, and we can see data only of the 3rd option that was selected.
Basically the `postdata.filters` is getting cleared of the 2 options selected previously.(as commented in the code below)
The MultiSelect UI Widget works properly with jqGrid 4.6 that I was using before.

The search values are created dynamically, for the columns that I require,
```JavaScript
var buildSearchSelect = function(uniqueNames) {
	var values = "";
	$.each(uniqueNames, function() {
		values += this + ":" + this + ";" ;
	});
	values = values.slice(0, -1);
	return values;
},
setSearchSelect = function(columnName, gridName) {
	jQuery(gridName).jqGrid('setColProp', columnName, {
		stype: 'select',
		searchoptions: {
			value: buildSearchSelect(getUniqueNames(columnName, gridName)),
			sopt: ['eq'],
			attr: { multiple: 'multiple', size: 4 },
			dataInit: dataInitMultiselect
		}
	});
},
myDefaultSearch = "cn",
getColumnIndexByName = function(columnName) {
	var cm = $(this).jqGrid('getGridParam', 'colModel'), i, l = cm.length;
	for (i = 0; i < l; i += 1) {
		if (cm[i].name === columnName) {
			return i; // return the index
		}
	}
	return -1;
};
```
The searching filter is constructed like this,
```JavaScript
var modifySearchingFilter = function(separator) {
	var i, l, rules, rule, parts, j, group, str, iCol, cmi, cm = this.p.colModel,
		filters = this.p.postData.filters === undefined ? null : $.parseJSON(this.p.postData.filters); 

//_On debugging at this point, filters value does not have selected option 1 and 2, when 3rd option is //selected. It has just the value of 3rd option_
	if (filters && filters.rules !== undefined && filters.rules.length > 0) {
		rules = filters.rules;
		for (i = 0; i < rules.length; i++) {
			rule = rules[i];
			iCol = getColumnIndexByName.call(this, rule.field);
			cmi = cm[iCol];
			if (iCol >= 0 &&
					((cmi.searchoptions === undefined || cmi.searchoptions.sopt === undefined)
						&& (rule.op === myDefaultSearch)) ||
					(typeof (cmi.searchoptions) === "object" &&
						$.isArray(cmi.searchoptions.sopt) &&
						cmi.searchoptions.sopt[0] === rule.op)) {
				// make modifications only for the 'contains' operation
				parts = rule.data.split(separator);
				if (parts.length > 1) {
					if (filters.groups === undefined) {
						filters.groups = [];
					}
					group = {
						groupOp: 'OR',
						groups: [],
						rules: []
					};
					filters.groups.push(group);
					for (j = 0, l = parts.length; j < l; j++) {
						str = parts[j];
						if (str) {
							// skip empty '', which exist in case of two separaters of once
							group.rules.push({
								data: parts[j],
								op: rule.op,
								field: rule.field
							});
						}
					}
					rules.splice(i, 1);
					i--; // to skip i++
				}
			}
		}
		this.p.postData.filters = JSON.stringify(filters);
	}
}
```
I set the multiselect filter for a column on loadComplete like,
```JavaScript
setSearchSelect('MY_COL', gridName);
```
I call `modifySearchingFilter` in `beforeRequest`

`dataInitMultiselect` is defined as,
```JavaScript
var dataInitMultiselect = function(elem) {
	setTimeout(function() {
		var $elem = $(elem), id = elem.id,
			inToolbar = typeof id === "string" && id.substr(0, 3) === "gs_",
			options = {
				selectedList: 2,
				height: "auto",
				checkAllText: "all",
				uncheckAllText: "no",
				noneSelectedText: "Any",
				open: function() {
					var $menu = $(".ui-multiselect-menu:visible");
					$menu.width("auto");
					$menu.find(">ul").css("maxHeight", "200px");
					return;
				}
			},
			$options = $elem.find("option");
		if ($options.length > 0 && $options[0].selected) {
			$options[0].selected = false; // unselect the first selected option

		}
		if (inToolbar) {
			options.minWidth = 'auto';
		}
		$elem.multiselect(options);
		$elem.siblings('button.ui-multiselect').css({
			width: inToolbar ? "98%" : "100%",
			marginTop: "1px",
			marginBottom: "1px",
			paddingTop: "3px"
		});
	}, 50);
};
```
But now I am using the free-jqgrid 4.13.3 for its additional cool features and would really like this issue resolved.

Can you please confirm.

Thanks.

**Comment by OlegKi** [comment]
@AshK14: I reformatted code fragments, which you posted. I hope it's all stay like in your code. In general I see that some parts of the code come from old jqGrid and the code can be reduced. I think that you have small compatibility problem.

It would be helpful if prepare the demo (in JSFiddle for example), which demonstrates the problem. I will modify the demo to demonstrate how the code could work with MultiSelect UI Widget.

I use myself MultiSelect UI Widget in one project and it works with free jqGrid 4.13.3. The only problem is the adjustment of your current code to make it working. The demo would be very helpful here.

**Comment by AshK14** [comment]
Hi Oleg,

I have prepared a small demo on JSFiddle to indicate this issue that I am facing.
In the fiddle, you can see that the multiselect widget is not filtering as it should and it also clears the text on the filtertoolbar in other columns when selecting more than one option in the widget.

The URL for JSFiddle,
https://jsfiddle.net/AshK14/aa5b9wn7/4/

Please let me know your thoughts on how to resolve this.

Thank You.

**Comment by OlegKi** [comment]
Thank you for the demo. I see the problem. The main reason of the problem is modification of `postData.filters` which makes some problems with new possibility: restoring the filter of the searching toolbar based on `postData.filters`. The filter generated by `modifySearchingFilter` can't full be displayed in the filter toolbar because it used multiple groups. The simplest way to fix the problem would be in the usage of `loadFilterDefaults: false` option of `filterToolbar` method. See the demo https://jsfiddle.net/aa5b9wn7/5/ which seems correctly to work.

The better way would be more deep modification of the old code, which I created for jqGrid 4.3.1-4.6 (see [the answer](http://stackoverflow.com/a/12875703/315935) and [this one](http://stackoverflow.com/a/26949859/315935)). Free jqGrid supports [custom filtering searching operation](https://github.com/free-jqgrid/jqGrid/wiki/Custom-filtering-searching-Operation), which allows to implement multiselect without modification of `postData.filters`. I will examine the problem more and will post you additional information later.

**Comment by OlegKi** [comment]
I modified the demo to the following: https://jsfiddle.net/OlegKi/aa5b9wn7/6/
It uses now **the latest source of free jqGrid** (from GitHub), where I added support of IN operation (`sopt: ["in"]`). You can see that the code is essentially reduced and everything seems to work. Please test it and report any problems which you could find. Finally, if all problems will be solved, please close the issue.

**Comment by AshK14** [comment]
Hi Oleg,

I simplified the code as per your suggestions and I am now using the latest source code from GitHub.
The issue is resolved and it works great!

Can you please confirm whether this change will be included in the next release of Free-jqgrid so that I can use the appropriate code in the future.

As always your answers have helped me in understanding jqgrid better.

Thanks!

**Comment by OlegKi** [comment]
You are welcome! The current code on the GitHub is the origin for the next version, which will be published. Probably I'll publish the code as the version 4.13.4 next days.

**Comment by OlegKi** [comment]
I published the version 4.13.4 today as I promised you before. It's available already from both CDNs ([jsDelivr](http://www.jsdelivr.com/#!free-jqgrid) and [cdnjs](https://cdnjs.com/libraries/free-jqgrid)) and of cause from npm, bower und NuGet.

**Comment by AshK14** [comment]
Thank You!
